### ISSUE TITLE
Option validation

### ISSUE THREAD

**Issue by platinumazure** [eslint/espree#384]
It could be useful to add option validation to espree and make sure that incoming options match a schema.

Are there any downsides? Do we want to allow variadic options? Should we make this validation itself opt-in?

(Inspired by eslint/eslint#10479.)

**Comment by not-an-aardvark** [comment]
One potential downside is that ESLint occasionally passes new options to all parsers to allow for feature detection (e.g. [`eslintScopeManager`](https://eslint.org/docs/developer-guide/working-with-plugins#working-with-custom-parsers) and [`eslintVisitorKeys`](https://eslint.org/docs/developer-guide/working-with-plugins#working-with-custom-parsers)). If we throw an error for unrecognized option, then we might end up with unexpected crashes when ESLint adds a new option to pass to parsers.

**Comment by platinumazure** [comment]
@not-an-aardvark Good call, I didn't think of that.

Maybe we could have a solution that allows some option wildcards (e.g., if the key starts with "eslint") but otherwise raise an error on unknown options?

**Comment by Aladdin-ADD** [comment]
Take a step back, we can simply check the option types, not values. e.g.
```js
{
ecmaVersion: "2018"
}
```

we can give a warning like `ecmaVersion should be a number, but you are passing "2018", do you mean 2018?`

**Comment by tapasweni-pathak** [comment]
:wave: @platinumazure @not-an-aardvark want to contribute to eslint codebase, working on a pr for this, do you think it would be good to implement a schema like [this](https://github.com/eslint/eslint/blob/master/conf/config-schema.js)? What specific options should be present in schema, what should be the error messages? great if you could add some more details in this issue.

**Comment by kaicataldo** [comment]
@tapasweni-pathak Sorry we missed this. I think we need to discuss this change some more, but we'd love your contribution if you're still interested!

@not-an-aardvark @platinumazure Another option could be to only validate keys we know (and have documented), rather than introducing additional complexity around special key name patterns.

**Comment by not-an-aardvark** [comment]
To clarify, do you mean we would validate the values of known options (e.g. making sure `ecmaVersion` is a number rather than a string) but we would continue to ignore unknown options? That sounds fine to me -- it seems like a clear improvement over what we do now.

**Comment by kaicataldo** [comment]
That's correct.

**Comment by tapasweni-pathak** [comment]
@kaicataldo @not-an-aardvark thanks for updating the ticket, picking this ticket, i'll ping in gitter for any other questions.

**Comment by nzakas** [comment]
Does anyone want to champion this and write up a full proposal? It's been open for several months so we should make a decision on whether or not to move forward.

**Comment by platinumazure** [comment]
I don't want to pursue this at this point. Happy to close in a couple of days if nobody wants to take this over before then.

**Comment by not-an-aardvark** [comment]
I'm not sure what a "full proposal" would entail here since the proposed change seems to be quite small. For the sake of iteration (and to illuminate any complexities that I might be missing), what questions would be left unanswered by following one-sentence proposal?

> Espree should throw an error when passed an `ecmaVersion` value which is neither `undefined` nor a number.

**Comment by nzakas** [comment]
In this case, a full proposal means, "can someone explain what everyone has agreed to?" In an issue with back and forth comments asking questions, the true proposal can get lost (part of why we have RFCs now).

If everyone agrees that your one sentence summarizes the proposal, then we can move forward.

**Comment by not-an-aardvark** [comment]
This proposal was accepted in [yesterday's TSC meeting](https://github.com/eslint/tsc-meetings/blob/137fbe2be55499cd75f28a008900803078d22dfd/notes/2019/2019-03-14.md#validate-ecmaversion-and-sourcetype-in-espree).
